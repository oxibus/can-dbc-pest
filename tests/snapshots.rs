#![cfg(feature = "encoding")]

use std::borrow::Cow;
use std::fs;
use std::path::Path;

use can_dbc_pest::{decode_cp1252, DbcParser, Rule};
use insta::{assert_debug_snapshot, with_settings};
use pest::Parser;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [
        (
            "tests/shared-test-files/dbc-cantools",
            "dbc-cantools",
            true,
            true,
        ),
        ("tests/shared-test-files/canpy", "canpy", true, true),
        ("tests/opendbc/opendbc/dbc", "opendbc", false, false),
    ];
    for (path, file_name, snapshot, cp1251) in test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path, snapshot, cp1251);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>, snapshot: bool, cp1251: bool) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(path, snapshot, cp1251);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: impl AsRef<Path>, snapshot: bool, cp1251: bool) {
    let path = path.as_ref();
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    match DbcParser::parse(Rule::file, &buffer) {
        Ok(pairs) => {
            if std::env::var("SKIP_INSTA").is_err()
                && (snapshot || std::env::var("FORCE_INSTA").is_ok())
            {
                assert_debug_snapshot!(file_name, pairs);
            }
        }
        Err(e) => panic!("Failed to parse {file_name}.dbc: {e:#?}"),
    }
}
