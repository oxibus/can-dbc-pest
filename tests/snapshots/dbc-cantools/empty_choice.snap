---
source: tests/snapshots.rs
---
[
    Pair {
        rule: file,
        span: Span {
            str: "VERSION \"\"\n\n\nNS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\nBS_:\n\nBU_: tx_node\nVAL_TABLE_ example_choice 255 \"Not available\" 254 \"Error\" ;\n\n\nBO_ 10 example_message: 3 tx_node\n SG_ no_choice : 16|8@1- (1,0) [0|0] \"\" Vector__XXX\n SG_ empty_choice : 8|8@1- (1,0) [0|0] \"\" Vector__XXX\n SG_ non_empty_choice : 0|8@1- (1,0) [0|0] \"\" Vector__XXX\n\n\n\nBA_DEF_  \"BusType\" STRING ;\nBA_DEF_DEF_  \"BusType\" \"CAN\";\nVAL_ 10 empty_choice ;\nVAL_ 10 non_empty_choice 255 \"Not available\" 254 \"Error\" ;\n\n",
            start: 0,
            end: 760,
        },
        inner: [
            Pair {
                rule: version,
                span: Span {
                    str: "VERSION \"\"\n",
                    start: 0,
                    end: 11,
                },
                inner: [
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"\"",
                            start: 8,
                            end: 10,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "",
                                    start: 9,
                                    end: 9,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: new_symbols,
                span: Span {
                    str: "NS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\n",
                    start: 13,
                    end: 337,
                },
                inner: [
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "NS_DESC_",
                            start: 21,
                            end: 29,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CM_",
                            start: 31,
                            end: 34,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_",
                            start: 36,
                            end: 43,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_",
                            start: 45,
                            end: 48,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "VAL_",
                            start: 50,
                            end: 54,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CAT_DEF_",
                            start: 56,
                            end: 64,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CAT_",
                            start: 66,
                            end: 70,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "FILTER",
                            start: 72,
                            end: 78,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_DEF_",
                            start: 80,
                            end: 91,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "EV_DATA_",
                            start: 93,
                            end: 101,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "ENVVAR_DATA_",
                            start: 103,
                            end: 115,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SGTYPE_",
                            start: 117,
                            end: 124,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SGTYPE_VAL_",
                            start: 126,
                            end: 137,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_SGTYPE_",
                            start: 139,
                            end: 153,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_SGTYPE_",
                            start: 155,
                            end: 165,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_TYPE_REF_",
                            start: 167,
                            end: 180,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "VAL_TABLE_",
                            start: 182,
                            end: 192,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_GROUP_",
                            start: 194,
                            end: 204,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_VALTYPE_",
                            start: 206,
                            end: 218,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIGTYPE_VALTYPE_",
                            start: 220,
                            end: 236,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BO_TX_BU_",
                            start: 238,
                            end: 247,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_REL_",
                            start: 249,
                            end: 260,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_REL_",
                            start: 262,
                            end: 269,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_DEF_REL_",
                            start: 271,
                            end: 286,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_SG_REL_",
                            start: 288,
                            end: 298,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_EV_REL_",
                            start: 300,
                            end: 310,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_BO_REL_",
                            start: 312,
                            end: 322,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SG_MUL_VAL_",
                            start: 324,
                            end: 335,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: bit_timing,
                span: Span {
                    str: "BS_:\n",
                    start: 337,
                    end: 342,
                },
                inner: [],
            },
            Pair {
                rule: nodes,
                span: Span {
                    str: "BU_: tx_node\n",
                    start: 343,
                    end: 356,
                },
                inner: [
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "tx_node",
                            start: 348,
                            end: 355,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: value_table,
                span: Span {
                    str: "VAL_TABLE_ example_choice 255 \"Not available\" 254 \"Error\" ;",
                    start: 356,
                    end: 415,
                },
                inner: [
                    Pair {
                        rule: ident_ext,
                        span: Span {
                            str: "example_choice",
                            start: 367,
                            end: 381,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "255",
                            start: 382,
                            end: 385,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"Not available\"",
                            start: 386,
                            end: 401,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "Not available",
                                    start: 387,
                                    end: 400,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "254",
                            start: 402,
                            end: 405,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"Error\"",
                            start: 406,
                            end: 413,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "Error",
                                    start: 407,
                                    end: 412,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: message,
                span: Span {
                    str: "BO_ 10 example_message: 3 tx_node",
                    start: 418,
                    end: 451,
                },
                inner: [
                    Pair {
                        rule: can_id,
                        span: Span {
                            str: "10",
                            start: 422,
                            end: 424,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident_ext,
                        span: Span {
                            str: "example_message",
                            start: 425,
                            end: 440,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "3",
                            start: 442,
                            end: 443,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "tx_node",
                            start: 444,
                            end: 451,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal,
                span: Span {
                    str: "SG_ no_choice : 16|8@1- (1,0) [0|0] \"\" Vector__XXX",
                    start: 453,
                    end: 503,
                },
                inner: [
                    Pair {
                        rule: ident_ext,
                        span: Span {
                            str: "no_choice",
                            start: 457,
                            end: 466,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "16",
                            start: 469,
                            end: 471,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "8",
                            start: 472,
                            end: 473,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "1",
                            start: 474,
                            end: 475,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "1",
                            start: 478,
                            end: 479,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 480,
                            end: 481,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 484,
                            end: 485,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 486,
                            end: 487,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"\"",
                            start: 489,
                            end: 491,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "",
                                    start: 490,
                                    end: 490,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "Vector__XXX",
                            start: 492,
                            end: 503,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal,
                span: Span {
                    str: "SG_ empty_choice : 8|8@1- (1,0) [0|0] \"\" Vector__XXX",
                    start: 505,
                    end: 557,
                },
                inner: [
                    Pair {
                        rule: ident_ext,
                        span: Span {
                            str: "empty_choice",
                            start: 509,
                            end: 521,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "8",
                            start: 524,
                            end: 525,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "8",
                            start: 526,
                            end: 527,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "1",
                            start: 528,
                            end: 529,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "1",
                            start: 532,
                            end: 533,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 534,
                            end: 535,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 538,
                            end: 539,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 540,
                            end: 541,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"\"",
                            start: 543,
                            end: 545,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "",
                                    start: 544,
                                    end: 544,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "Vector__XXX",
                            start: 546,
                            end: 557,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal,
                span: Span {
                    str: "SG_ non_empty_choice : 0|8@1- (1,0) [0|0] \"\" Vector__XXX",
                    start: 559,
                    end: 615,
                },
                inner: [
                    Pair {
                        rule: ident_ext,
                        span: Span {
                            str: "non_empty_choice",
                            start: 563,
                            end: 579,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "0",
                            start: 582,
                            end: 583,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "8",
                            start: 584,
                            end: 585,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "1",
                            start: 586,
                            end: 587,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "1",
                            start: 590,
                            end: 591,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 592,
                            end: 593,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 596,
                            end: 597,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 598,
                            end: 599,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"\"",
                            start: 601,
                            end: 603,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "",
                                    start: 602,
                                    end: 602,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "Vector__XXX",
                            start: 604,
                            end: 615,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: attr_def,
                span: Span {
                    str: "BA_DEF_  \"BusType\" STRING ;",
                    start: 619,
                    end: 646,
                },
                inner: [
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"BusType\"",
                            start: 628,
                            end: 637,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "BusType",
                                    start: 629,
                                    end: 636,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: ba_def_def,
                span: Span {
                    str: "BA_DEF_DEF_  \"BusType\" \"CAN\";",
                    start: 647,
                    end: 676,
                },
                inner: [
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"BusType\"",
                            start: 660,
                            end: 669,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "BusType",
                                    start: 661,
                                    end: 668,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"CAN\"",
                            start: 670,
                            end: 675,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "CAN",
                                    start: 671,
                                    end: 674,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: value_table_def,
                span: Span {
                    str: "VAL_ 10 empty_choice ;",
                    start: 677,
                    end: 699,
                },
                inner: [
                    Pair {
                        rule: can_id,
                        span: Span {
                            str: "10",
                            start: 682,
                            end: 684,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "empty_choice",
                            start: 685,
                            end: 697,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: value_table_def,
                span: Span {
                    str: "VAL_ 10 non_empty_choice 255 \"Not available\" 254 \"Error\" ;",
                    start: 700,
                    end: 758,
                },
                inner: [
                    Pair {
                        rule: can_id,
                        span: Span {
                            str: "10",
                            start: 705,
                            end: 707,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "non_empty_choice",
                            start: 708,
                            end: 724,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "255",
                            start: 725,
                            end: 728,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"Not available\"",
                            start: 729,
                            end: 744,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "Not available",
                                    start: 730,
                                    end: 743,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "254",
                            start: 745,
                            end: 748,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"Error\"",
                            start: 749,
                            end: 756,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "Error",
                                    start: 750,
                                    end: 755,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: EOI,
                span: Span {
                    str: "",
                    start: 760,
                    end: 760,
                },
                inner: [],
            },
        ],
    },
]
